demos/link-prediction/random-walks/ctdne-link-prediction.ipynb

```diff
@@ -0,0 +1,725 @@
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Number of edges in graph: 37929\n",
+      "Number of edges in training set: 9482\n",
+      "Number of edges in test set: 3161\n"
+     ]
+    }
+   ],
+   "source": [
+    "# subset of edges to split\n",
+    "train_subset = 0.25\n",
+    "test_subset = 0.25\n",
+    "\n",
+    "# number of edges to be kept in the graph\n",
+    "num_edges_graph = int(len(edges) * (1 - train_subset))\n",
+    "\n",
+    "# keep older edges in graph, and predict more recent edges\n",
+    "edges_graph = edges[:num_edges_graph]\n",
+    "edges_other = edges[num_edges_graph:]\n",
+    "\n",
+    "# split recent edges further to train and test sets\n",
+    "edges_train, edges_test = train_test_split(edges_other, test_size=test_subset)\n",
+    "\n",
+    "print(\n",
+    "    f\"Number of edges in graph: {len(edges_graph)}\\n\"\n",
+    "    f\"Number of edges in training set: {len(edges_train)}\\n\"\n",
+    "    f\"Number of edges in test set: {len(edges_test)}\"\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Create a StellarGraph\n",
+    "\n",
+    "Now we can use the edges we've reserved for the graph to create an instance of the StellarGraph class"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "graph = StellarGraph(\n",
+    "    nodes=pd.DataFrame(index=full_graph.nodes()),\n",
+    "    edges=edges_graph,\n",
+    "    source_column=\"source\",\n",
+    "    target_column=\"target\",\n",
+    "    edge_weight_column=\"time\",\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Create link examples for training and testing\n",
+    "\n",
+    "The edges that we've reserved for the train/test sets become examples of positive links. \n",
+    "\n",
+    "We also need to randomly generate some examples of negative links in order to train our classifier."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def create_link_examples(g, edges):\n",
+    "    pos = [(src, tgt) for (src, tgt) in edges[[\"source\", \"target\"]].values]\n",
+    "    neg = sample_negative_examples(g, pos)\n",
+    "    return pos, neg\n",
+    "\n",
+    "\n",
+    "def sample_negative_examples(g, positive_examples):\n",
+    "    possible_edges_all = {\n",
+    "        (src, tgt) for src in g.nodes() for tgt in g.nodes() if src != tgt\n",
+    "    }\n",
+    "    possible_edges_neg = list(\n",
+    "        possible_edges_all.difference(\n",
+    "            {e for (src, tgt) in positive_examples for e in [(src, tgt), (tgt, src)]}\n",
+    "        )\n",
+    "    )\n",
+    "    return random.choices(possible_edges_neg, k=len(positive_examples))\n",
```

Comment:
Do you mean to sample with replacement?

Currently this is extremely likely to have overlaps. There's 151&times;150 = 22650 non-self loop edges in a graph with 151 nodes, and out of the 50572 (total edges) &times; 0.25 (graph vs. train/test split) &times; 0.75 (train split) = 9482 positive train edges, it seems there's usually ~1300 unique undirected edges. This means this `choices` is sampling with replacement N=9482 examples from K=~21350 possible edges, which makes a multinomial distribution with probability vector (1/K, 1/K, ..., 1/K) (length K) and sample size N. We can work out the expected number of the K unique values that are sampled, by subtracting the number that *aren't* sampled from the total number K: `K - K * (1 - 1/K)**N = 7658.1` (https://stats.stackexchange.com/questions/420643/expected-number-of-zeros-in-multinomial-vector). That is, this is, on average, getting only 7658 unique negative edges across the 9482 elements of list here. (That theoretical average matches pretty closely to values I see in practice, across a few runs: 7658, 7643, 7614, 7700.) The probability of no duplicates is rather small: ~`(1 - 1/K)**(N*(N-1)/2)` &approx; 4&times;10<sup>-913</sup> (https://en.wikipedia.org/wiki/Birthday_problem#Cast_as_a_collision_problem).

I suspect this may be incorrect, which case, maybe this should be done without replacement, e.g. via `random.sample`?

```suggestion
    "    return random.sample(possible_edges_neg, k=len(positive_examples))\n",
```

Reply:
Thanks, yes I think without replacement makes sense, and perhaps explains the high scores, I'll give this a go